### DISCUSSION TITLE
Injecting a javascript file into every page of the Airflow UI

### DISCUSSION THREAD

**Discussion by nkoep** [apache/airflow#25811] [General]
Hey everyone,

I was wondering if there was a way (e.g. through a plugin) to load a javascript file into every airflow page to extend the UI in certain places. I can see how static script files could be included in custom views added through a plugin, but I couldn't find anything about how to load an additional script file everywhere.

**Comment by barksdaleaz** [comment]
Hi @nkoep - were you able to find a solution for this? I'm currently having a very similar problem and nobody seems to know!

**Comment by nkoep** [comment]
Unfortunately not, sorry 🫤

**Comment by jdistler** [comment]
Sharing my solution here in case it helps anyone. Airflow lets you modify the flask application directly, see here - 
https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html#configuring-flask-application-for-airflow-webserver

You can do all kinds of hacky tricks once you control flask like this. For example in my `webserver_config.py` I did this - 

```
@app.after_request
def after_request(response):
    # Adding script tag that will be injected in every page
    script = """
    <script>
        console.log("BOOYAH");
    </script>
    """

    if response.content_type.startswith('text/html'):
        # If the response data is gzipped.
        gzip_header = response.headers.get('Content-Encoding')
        if gzip_header and gzip_header.lower() == 'gzip':
            decompressed_data = gzip.decompress(response.get_data())
            # Add the script into the decompressed HTML
            decompressed_html = decompressed_data.decode().replace("</body>", f"{script}</body>")
            # Compress the html again
            bio = BytesIO()
            with gzip.GzipFile(fileobj=bio, mode='w') as gzip_file:
                gzip_file.write(decompressed_html.encode())
            response.set_data(bio.getvalue())
        else:  # If the response data is not gzipped.
            response.set_data(
                response.get_data(as_text=True).replace("</body>", f"{script}</body>")
            )
        print('Script injected.')
    return response
```

And then in every page I got this script appended and does whatever logic I need. This explicitly hijacks the response from every request and modifies the body to append my script.
